include/deal.II/lac/diagonal_matrix.h

```diff
@@ -0,0 +1,290 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2016 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__diagonal_matrix_h
+#define dealii__diagonal_matrix_h
+
+
+#include <deal.II/lac/vector.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * This class represents a diagonal matrix based on a vector.
+ *
+ * @author Martin Kronbichler, 2016
+ */
+template <typename VectorType=Vector<double> >
+class DiagonalMatrix : public Subscriptor
+{
+public:
+  typedef typename VectorType::value_type value_type;
+  typedef typename VectorType::size_type size_type;
+
+  /**
+   * Constructor.
+   */
+  DiagonalMatrix();
+
+  /**
+   * Initialize with a given vector by copying the content of the vector
+   * @p vec.
+   */
+  void reinit (const VectorType &vec);
+
+  /**
+   * Returns a reference to the underlying vector for manipulation of the
+   * entries on the diagonal.
+   */
+  VectorType &get_vector();
+
+  /**
+   * Returns a read-only reference to the underlying vector.
+   */
+  const VectorType &get_vector() const;
+
+  /**
+   * Number of rows of this matrix. To remember: this matrix is an <i>m x
+   * m</i>-matrix.
+   */
+  size_type m () const;
+
+  /**
+   * Number of columns of this matrix. To remember: this matrix is an <i>n x
+   * n</i>-matrix.
+   */
+  size_type n () const;
+
+  /**
+   * Read-only access to a value. This is restricted to the case where
+   * <i>i==j</i>.
+   *
+   * If the underlying vector is distributed with MPI, not all of the indices
+   * <i>i</i> might actually be accessible. Refer to the method
+   * <code>get_vector().locally_owned_elements()</code> for the entries that
+   * actually are accessible.
+   */
+  value_type operator()(size_type i, size_type j) const;
+
+  /**
+   * Read-write access to a value. This is restricted to the case where
+   * <i>i==j</i>.
+   *
+   * If the underlying vector is distributed with MPI, not all of the indices
+   * <i>i</i> might actually be accessible. Refer to the method
+   * <code>get_vector().locally_owned_elements()</code> for the entries that
+   * actually are accessible.
+   */
+  value_type &operator()(size_type i, size_type j);
+
+  /**
+   * Add an array of values given by <tt>values</tt> in the given global
+   * matrix row at columns specified by col_indices. Due to the storage of
+   * this matrix, entries are only added to the diagonal of the matrix. All
+   * other entries are ignored and no exception is thrown.
```

Comment:
Exactly, in the context of `ConstraintMatrix::distribute_local_to_global` where this might be used, we do not want to write separate code paths. Even if a user only passes a diagonal matrix to that function, constraints will create non-diagonal contributions nonetheless, so we are forced to skip non-diagonal entries rather than assert at this stage. Unless we see a good alternative to this, I'd keep the semantics like this. (Either I or @davydden will add that class to ConstraintMatrix in a later PR.)